src/Raven.Server/Documents/PeriodicBackup/Aws/RavenAwsS3Client.cs

```diff
@@ -500,6 +502,98 @@ public async Task DeleteBucket()
             }
         }
 
+        public async Task<IEnumerable<FileInfoDetails>> ListObjects(string prefix, string delimiter, bool listFolders, string continuationToken = null)
+        {
+            var url = $"{GetUrl()}/?list-type=2";
+            if (prefix != null)
+                url += $"&prefix={prefix}";
+
+            if (delimiter != null)
+                url += $"&delimiter={delimiter}";
+
+            if (continuationToken != null)
+                url += $"&continuation-token={Uri.EscapeDataString(continuationToken)}";
+
+            var now = SystemTime.UtcNow;
+
+            var requestMessage = new HttpRequestMessage(HttpMethods.Get, url);
+            UpdateHeaders(requestMessage.Headers, now, null);
+
+            var headers = ConvertToHeaders(requestMessage.Headers);
+
+            var client = GetClient();
+            client.DefaultRequestHeaders.Authorization = CalculateAuthorizationHeaderValue(HttpMethods.Get, url, now, headers);
+
+            var response = await client.SendAsync(requestMessage, CancellationToken);
+            if (response.StatusCode == HttpStatusCode.NotFound)
+                return new List<FileInfoDetails>();
+
+            if (response.IsSuccessStatusCode == false)
+                throw StorageException.FromResponseMessage(response);
+
+            var responseStream = await response.Content.ReadAsStreamAsync();
+            var listBucketResult = XDocument.Load(responseStream);
+            var ns = listBucketResult.Root.Name.Namespace;
+            var result = GetResult();
+
+            var isTruncated = listBucketResult.Root.Element(ns + "IsTruncated").Value;
+            if (isTruncated == "true")
```

Comment:
This worries me. 
What happens if we have a folder that has thousands of backups?
Won't this cause us to scan it each and every time?

Reply:
Yes. We can optimize this by taking all the files in all the folders in a single call and parse it to folders locally (there are no "folders" in AWS).

Reply:
Is this optimization critical? I think we can add it later on? (create an issue for that)

@ayende what do you think?

Reply:
It can wait for a later issue.